compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/CodeAnalyzer.java

```diff
@@ -2462,6 +2476,114 @@ public void visit(BLangSimpleVariable varNode) {
+                return validateContext(node.parent);
+            default:
+                return false;
+        }
+    }
+
+    private void validateInferredArrays(BLangNode node, BLangExpression expr) {
+        switch (node.getKind()) {
+            case ARRAY_TYPE:
+                BLangArrayType arrayType = (BLangArrayType) node;
+                if (isInferredArray(arrayType.sizes)) {
+                    if (expr != null && expr.getKind() == NodeKind.LIST_CONSTRUCTOR_EXPR &&
+                            validateInitializer(arrayType.sizes, arrayType.sizes.size() - 2,
+                                    (BLangListConstructorExpr) expr)) {
+                        arrayType.isValidInferredArray = true;
+                    }
+                }
+                return;
+            case INTERSECTION_TYPE_NODE:
+                BLangIntersectionTypeNode intersectionTypeNode = (BLangIntersectionTypeNode) node;
+                for (BLangType member : intersectionTypeNode.constituentTypeNodes) {
+                    validateInferredArrays(member, expr);
+                }
+                return;
+            case UNION_TYPE_NODE:
+                BLangUnionTypeNode unionTypeNode = (BLangUnionTypeNode) node;
+                for (BLangType member : unionTypeNode.memberTypeNodes) {
+                    validateInferredArrays(member, expr);
+                }
+                return;
+            default:
+                return;
+        }
+    }
+
+    private void getPositionsOfInferredArrays(BLangNode node, List<Location> posArray) {
+        switch (node.getKind()) {
+            case ARRAY_TYPE:
+                BLangArrayType arrayType = (BLangArrayType) node;
+                if (isInferredArray(arrayType.sizes)) {
+                    posArray.add(arrayType.pos);
+                }
+                return;
+            case INTERSECTION_TYPE_NODE:
+                BLangIntersectionTypeNode intersectionTypeNode = (BLangIntersectionTypeNode) node;
+                for (BLangType member : intersectionTypeNode.constituentTypeNodes) {
+                    getPositionsOfInferredArrays(member, posArray);
+                }
+                return;
+            case UNION_TYPE_NODE:
+                BLangUnionTypeNode unionTypeNode = (BLangUnionTypeNode) node;
+                for (BLangType member : unionTypeNode.memberTypeNodes) {
+                    getPositionsOfInferredArrays(member, posArray);
+                }
+                return;
+            default:
+                return;
+        }
+    }
+
+    private boolean validateInitializer(List<BLangExpression> indexSizes, int index, BLangListConstructorExpr expr) {
+        if (index < 0) {
+            return true;
+        }
+
+        boolean isInferred = false;
+        if ((indexSizes.get(index).getKind() == LITERAL) &&
+                ((BLangLiteral) indexSizes.get(index)).value.equals(Constants.INFERRED_ARRAY_INDICATOR)) {
+            isInferred = true;
+        }
```

Comment:
Can we inline this by removing the condition?

```
boolean isInferred = (indexSizes.get(index).getKind() == LITERAL) &&
                ((BLangLiteral) indexSizes.get(index)).value.equals(Constants.INFERRED_ARRAY_INDICATOR);
```

Reply:
Yeah. Will update. Thanks.